src/app/interaction-model/MetadataTypes.h

```diff
@@ -0,0 +1,159 @@
+    BitFlags<ClusterQualityFlags> flags;
+};
+
+struct ClusterEntry
+{
+    ConcreteClusterPath path;
+    ClusterInfo info;
+
+    static ClusterEntry Invalid()
+    {
+        ClusterEntry result;
+        result.path = ConcreteClusterPath(kInvalidEndpointId, kInvalidClusterId);
+        return result;
+    }
+};
+
+enum class AttributeQualityFlags : uint32_t
+{
+    kListAttribute   = 0x0004, // This attribute is a list attribute
+    kFabricScoped    = 0x0008, // 'F' quality on attributes
+    kFabricSensitive = 0x0010, // 'S' quality on attributes
+    kChangesOmitted  = 0x0020, // `C` quality on attributes
+    kTimed           = 0x0040, // `T` quality on attributes (writes require timed interactions)
+};
+
+struct AttributeInfo
+{
+    BitFlags<AttributeQualityFlags> flags;
+
+    // read/write access will be missing if read/write is NOT allowed
+    std::optional<Access::Privilege> readPrivilege;  // generally defaults to View if readable
+    std::optional<Access::Privilege> writePrivilege; // generally defaults to Operate if writable
+};
+
+struct AttributeEntry
+{
+    ConcreteAttributePath path;
+    AttributeInfo info;
+
+    static AttributeEntry Invalid()
+    {
+        AttributeEntry result;
+        result.path = ConcreteAttributePath(kInvalidEndpointId, kInvalidClusterId, kInvalidAttributeId);
+        return result;
+    }
+};
+
+enum class CommandQualityFlags : uint32_t
+{
+    kFabricScoped    = 0x0001,
+    kFabricSensitive = 0x0002,
+    kTimed           = 0x0004, // `T` quality on commands
+};
+
+struct CommandInfo
+{
+    BitFlags<CommandQualityFlags> flags;
+    Access::Privilege invokePrivilege = Access::Privilege::kOperate;
+};
+
+struct CommandEntry
+{
+    ConcreteCommandPath path;
+    CommandInfo info;
+
+    static CommandEntry Invalid()
+    {
+        CommandEntry result;
+        result.path = ConcreteCommandPath(kInvalidEndpointId, kInvalidClusterId, kInvalidCommandId);
+        return result;
+    }
+};
+
+/// Provides metadata information for a data model
+///
+/// The data model can be viewed as a tree of endpoint/cluster/(attribute+commands+events)
+/// where each element can be iterated through independently.
+///
+/// Iteration rules:
+///   - kInvalidEndpointId will be returned when iteration ends (or generally kInvalid* for paths)
+///   - Global Attributes are NOT returned since they are implied
+///   - Any internal iteration errors are just logged (callers do not handle iteration CHIP_ERROR)
+///   - Iteration order is NOT guaranteed globally. Only the following is guaranteed:
+///     - Complete tree iteration (e.g. when iterating an endpoint, ALL clusters of that endpoint
+///       are returned, when iterating over a cluster, all attributes/commands are iterated over)
+///     - uniqueness and completeness (iterate over all possible distinct values as long as no
+///       internal structural changes occur)
+class DataModelMetadataTree
```

Comment:
converted to draft as there are significant changes here and I need to implement them. Will remove out of draft once implementation is ready.

Reply:
Updated.